`Team exercises`: Add csv file team import and provide better error messages

Hialus opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
Closes #4255 
Closes #4256 
In order to use external tools to create teams of students for an exercise, Artemis has an import teams function. This function however only supports json files. Tools like Moodle however export their teams as csv files. So it would be useful to be able to import these files directly.
Also, the error messages when the imported file was malformed were quite poor. Since these files can contains hundreds or thousands of entries it is important to at least tell the user which entry is responsible for the error.

### Description
<!-- Describe your changes in detail -->
For the import of csv files I simply changed the file import to also accept csv files which are then parsed with PapaParse. They accept the same keys as the json file requires, but also accept some additional keys like `Matrikelnummer` instead of `registrationNumber`. For example this will allow instructors to directly import files from Moodle (The format of which was provided in #1994)
For the error messages I changed the language keys to also include parameters that describe which entry is responsible for the error. The error messages now look similar to this:
> The team name 1Team XYZ in entry 101 does not start with a letter.

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->
Prerequisites:
- 1 Instructor
- A course with a couple of registered students
- 1 Exercise with teams enabled
- Some valid and invalid json files (You can create these by exporting teams from another exercise)
- Some valid and invalid csv files (You can create these by exporting teams from another exercise and convert the json file here: https://www.convertcsv.com/json-to-csv.htm)

1. Log in to Artemis
2. Navigate to the course management exercise page of the exercise that has teams enabled
3. Navigate to the teams page of this exercise
4. Click on Import Teams
5. Import the csv or json files
    - The valid files should import with no problem
    - The invalid files should provide proper error messages

### Review Progress
<!-- Each Pull Request should be reviewed by at least two other developers. The code as well as the functionality (= manual test) needs to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
<!-- Please add the test coverages for all changed files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan. -->
<!-- Please use the schema "Branches % | Lines %" -->
<!-- Lines are the main reference but a significantly lower branch percentage can indicate missing edge cases in the tests. -->
<!-- - ExerciseService.java: 85% | 77% -->
<!-- - programming-exercise.component.ts: 13% | 95% -->
File: Lines % | Branches %
teams-import-from-file-form.component.ts:  95.95% | 73.33%

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
No direct UI changes. Only the error messages are different.

mallgaier reviewed (approved):
Tested together with @lBruchmann and @Strohgelaender.  Import tested using JSON. Users with incorrect usernames are ignored and not imported.
If a team name does not start with a letter, a corresponding browser error message is displayed.

lBruchmann reviewed (dismissed):
Tested together with @Strohgelaender and @mallgaier. Tested on ts1. I followed the test steps. Importing valid and invalid (both .json and .csv files) works as expected. We have one small comment, see @Strohgelaender .

Strohgelaender reviewed (dismissed):
Tested in testing session with @lBruchmann @mallgaier on ts1. The import worked for both csv and json files and the enhanced error messaged were shown if invalid input was provided. But directly pasting the example provided in #1994 did not work, since it uses both `;` and `,` as a separator. We might have to recheck whether or not this is indeed the format that is provided by moodle, if yes the import stil won't work.

Hialus commented:
> Tested in testing session with @lBruchmann @mallgaier on ts1. The import worked for both csv and json files and the enhanced error messaged were shown if invalid input was provided. But directly pasting the example provided in #1994 did not work, since it uses both `;` and `,` as a separator. We might have to recheck whether or not this is indeed the format that is provided by moodle, if yes the import stil won't work.

If this is actually the format that Moodle uses the user will have to fix this before importing it.
While the tool used for parsing the CSV file (Papa Parse) tries to automatically detect the used separator, it does not support different separators in one file.

Garfield96 reviewed (changes requested):
Nice PR. Thanks for fixing my issues. Please update the UI and documentation to show the user which input formats are supported now.

Hialus commented:
I have removed the code duplication in the test cases and added the supported file types to the import dialog.
However, I could not find any documentation about the teams import feature, so I cannot update it. And it would be out of scope for this PR to add an entire documentation page about the teams feature.

julian-christl commented on src/test/javascript/spec/component/team/teams-import-from-file-form.component.spec.ts:
```diff
@@ -49,81 +44,94 @@ describe('TeamsImportFromFileFormComponent', () => {
+        let getElementStub: jest.SpyInstance;
         const element = document.createElement('input');
-        const control = { ...element, value: 'test' };
+        let control: any;
```
Please avoid `any`.

Malyuk-A commented on src/main/webapp/i18n/de/team.json:
```diff
@@ -41,8 +41,8 @@
             "teamName": {
-                "missingTeamName": "Alle Studierenden müssen einem Teamnamen zuordnet sein.",
-                "pattern": "Der Teamname muss mit einem Buchstaben beginnen."
+                "missingTeamName": "Studierende:r {{ studentName }} in Eintrag {{ entryNr }} ist kein Teamname zuordnet.",
```
```suggestion
                "missingTeamName": "Studierende:r {{ studentName }} in Eintrag {{ entryNr }} ist kein Teamname zugeordnet.",
```

Malyuk-A reviewed (commented):
small type, else approve

julian-christl commented on src/main/webapp/app/exercises/shared/team/teams-import-dialog/teams-import-from-file-form.component.ts:
```diff
@@ -73,11 +104,58 @@ export class TeamsImportFromFileFormComponent {
     convertTeams(importTeam: StudentWithTeam[]): Team[] {
         const teams: Team[] = [];
+        let entryNr = 1;
         importTeam.forEach((student) => {
```
```suggestion
        importTeam.forEach((student, index) => {
```
You can the remove `entryNr` and use `index + 1`

julian-christl reviewed (approved):
Code looks good now

Malyuk-A reviewed (approved):
Code looks good

daniels98it reviewed (approved):
No complaint on code 👍
